sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/BulkWriter.scala

```diff
@@ -0,0 +1,114 @@
+
+case class BulkWriter(container: CosmosAsyncContainer,
+                      writeConfig: CosmosWriteConfig) {
+
+  private val activeTasks = new AtomicInteger(0)
+  private val lock = new ReentrantLock
+  private val pendingTasksCompleted = lock.newCondition
+  private val errorCaptureFirstException = new AtomicReference[Throwable]()
+  private val bulkInputEmitter: EmitterProcessor[CosmosItemOperation] = EmitterProcessor.create
+
+  // TODO: moderakh implement retry
+  // TODO: moderakh handle 409
+
+  private val subscriptionDisposable: Disposable = {
+    val bulkOperationResponseFlux: SFlux[CosmosBulkOperationResponse[Object]] =
+      container.processBulkOperations[Object](bulkInputEmitter).asScala
+
+    bulkOperationResponseFlux.subscribe(
+      resp => {
+        resp.getBatchContext
+        if (resp.getException != null) {
+          // TODO: moderakh we should retry
+          captureIfFirstFailure(resp.getException)
+          cancelWork()
+        }
+        markTaskCompletion()
+      },
+      errorConsumer = Option.apply(
+        ex => {
+          // if there is any failure this closes the bulk.
+          // at this point bulk api doesn't allow any retrying
+          // we don't know the list of failed item-operations
+          // they only way to retry to keep a dictionary of pending operations outside
+          // so we know which operations failed and which ones can be retried.
+          // TODO: moderakh discuss the bulk API in the core SDK.
+          captureIfFirstFailure(ex)
+          cancelWork()
+          markTaskCompletion()
+        }
+      )
+    )
+  }
+
+  def scheduleWrite(partitionKeyValue: PartitionKey, objectNode: ObjectNode): Unit = {
+    val bulkItemOperation = writeConfig.itemWriteStrategy match {
+      case ItemWriteStrategy.ItemOverwrite =>
+        BulkOperations.getUpsertItemOperation(objectNode, partitionKeyValue)
+
+      // TODO moderakh add support for non upsert mode in bulk
+      case _ =>
+        throw new RuntimeException(s"${writeConfig.itemWriteStrategy} not supported")
+    }
+
+    activeTasks.incrementAndGet()
+    bulkInputEmitter.onNext(bulkItemOperation)
+  }
+
+  // the caller has to ensure that after invoking this method scheduleWrite doesn't get invoked
+  def flushAndClose(): Unit = {
+    bulkInputEmitter.onComplete()
+
+    // TODO: error handling, if there is any error and the subscription is cancelled
+    // the remaining tasks will not be processed hence we never reach 0 and the following will block
+    // once we do error handling we should think how to cover the scenario.
+
+    lock.lock()
+    try {
+      while (activeTasks.get() > 0) {
+        pendingTasksCompleted.await()
+      }
+
+    } finally {
+      lock.unlock()
+    }
+
+    // which error to report?
+    if (errorCaptureFirstException.get() != null) {
+      throw errorCaptureFirstException.get()
+    }
+  }
+
+  private def markTaskCompletion() : Unit = {
```

Comment:
Use `synchronized` unless we want to sync this method with flushAndClose?

Reply:
Since they have the same lock, what if flushAndClose is called first, wouldn't this deadlock? Because the task would never complete because the flushAndClose is holding the same lock while waiting for completion?

Reply:
This is Producer/Consumer pattern that one wakes up the other one for this you need a condition monitor and a lock.

we can't achieve this with synchronized.

please read the comment on `flushAndClose`